src/ldp/utils.py

```diff
@@ -110,3 +113,54 @@ def format_error_details(error: Exception) -> str:
             error_details += f"\nResponse body: {error.response.text}"
 
     return error_details
+
+
+def split_message_transitions(list_of_messages: list[Message]) -> list[list[Message]]:
+    """
+    Break down messages into transitions: [(system, user), (tool request, tool response, env state), ...].
+
+    Blocks end when either:
+    - an EnvStateMessage to ToolRequestMessage transition occurs
+    - a ToolResponseMessage to ToolRequestMessage transition occurs
+
+    Args:
+        list_of_messages: The list of messages to break down.
+
+    Returns:
+        A list of transitions blocks, corresponding to trajectory transition.
+    """
+    if not list_of_messages:
+        return []
+
+    filtered_messages: list[list[Message]] = []
+    i = 0
+
+    # First block: collect all system messages + user message
+    system_block: list[Message] = []
+    while i < len(list_of_messages) and list_of_messages[i].role in {"system", "user"}:
+        system_block.append(list_of_messages[i])
+        i += 1
```

Comment:
```suggestion
    for msg in list_of_messages:
        if msg.role not in {'system', 'user'}:
            break
        system_block.append(msg)
```

I found it harder to follow the index logic